src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9982,280 @@ struct C_Flush : public Context {
   }
 };
 
+struct C_SetDedupChunks : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  epoch_t last_peering_reset;
+  ceph_tid_t tid;
+  uint64_t offset;
+  
+  C_SetDedupChunks(PrimaryLogPG *p, hobject_t o, epoch_t lpr)
+    : pg(p), oid(o), last_peering_reset(lpr),
+      tid(0), offset(0)
+  {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    std::scoped_lock locker{*pg};
+    if (last_peering_reset != pg->get_last_peering_reset()) {
+      return;
+    }
+    pg->finish_set_dedup(oid, r, tid, offset);
+  }
+};
+
+
+int PrimaryLogPG::start_dedup(OpRequestRef op, ObjectContextRef obc)
+{
+  bufferlist bl;
+  uint64_t cur_off = 0;
+  const object_info_t& oi = obc->obs.oi;
+  const hobject_t& soid = oi.soid;
+
+  ceph_assert(obc->is_blocked());
+  if (oi.size == 0) {
+    // evicted 
+    return 0;
+  }
+
+  /*
+   * The operations to make dedup chunks are tracked by a ManifestOp.
+   * This op will be finished if all the operations are completed.
+   */
+  ManifestOpRef mop(std::make_shared<ManifestOp>(nullptr, 0));
+
+  while (cur_off < oi.size) {
+    // cdc
+    vector<pair<uint64_t, uint64_t>> chunks;
+    int r = do_cdc(oi, bl, chunks);
+    if (r < 0) {
+      return r;
+    }
+    if (!chunks.size()) {
+      break;
+    }
+
+    // get fingerprint 
+    for (auto p : chunks) {
+      bufferlist chunk;
+      chunk.substr_of(bl, p.first, p.second);
+      hobject_t target = get_fpoid_from_chunk(soid, chunk);
+      if (target == hobject_t()) {
+	dout(0) << " fingerprint oid is null " << dendl;
+	break;
+      }
+
+      // skip if the same content exits in prev snap at same offset
```

Comment:
I mentioned in another comment that I'm not sure what's supposed to happen if the object already has some references.  Let's assume that the operation is:
1) calculate new fingerprint and chunks
2) drop existing references
3) take new references

I think you need to use object_manifest_t::calc_refs_to_drop_on_removal to calculate the references you need to drop.  You can then use calf_refs_to_inc_on_set to get the set of references you need to add.  Those already correctly handle snapshots and should simplify your code.  Note, you need to defer dropping the references until after commit.

Reply:
If this doesn't match your intentions, let's talk about it here.

Reply:
Sorry, my bad.